### PR TITLE
BUG: loc setitem with missing integer slices does not raise

### LINKED ISSUE #26412: Using .loc.__setitem__ with integer slices does not raise
#### Code Sample

```python
>>> d = pd.DataFrame({'a': range(5), 'b': range(2,7), 'c': range(3,8)})
>>> d
   a  b  c
0  0  2  3
1  1  3  4
2  2  4  5
3  3  5  6
4  4  6  7
>>> d.loc[0:2, 0:2] = 42   # note that loc is used with an integer slice! should this raise?
>>> d  # this is somehow the expected result
    a   b  c
0  42  42  3
1  42  42  4
2  42  42  5
3   3   5  6
4   4   6  7
>>> d.iloc[0:2, 0:2]        # expected result
    a   b
0  42  42
1  42  42
>>> d.loc[0:2, 0:2]        # this is the same syntax as above, but raises
...
TypeError: cannot do slice indexing on <class 'pandas.core.indexes.base.Index'> with these indexers [0] of <class 'int'>
```
#### Problem description

I'd expect a symmetry in `d.loc.__getitem__` and `d.loc.__setitem__`. At the moment only __getitem__ raises if integer slices are passed. Usage of `.loc` was probably a mistake, instead `.iloc` should have been used. Even if it works correctly, it is confusing for users.

Probably this is simply a missing exception in the special case of `d.loc.__setitem__` being called with integer slices. 

See #13831 for a very similar problems with `pd.Series`.

#### Expected Output

Assuming `d.loc.__getitem__` should raise for integer slices (as it does), then also `d.loc.__setitem__` should raise:

```
>>> d.loc[0:2, 0:2]
...
SomException: int & loc don't go together
```

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.8.final.0
python-bits: 64
OS: Linux
OS-release: 4.18.0-20-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: 3.6.4
pip: 9.0.1
setuptools: 40.8.0
Cython: None
numpy: 1.16.2
scipy: 1.2.1
pyarrow: None
xarray: 0.11.3
IPython: 5.5.0
sphinx: 1.7.9
patsy: 0.4.1+dev
dateutil: 2.8.0
pytz: 2019.1
blosc: 1.5.1
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.5
feather: None
matplotlib: 2.2.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: 4.2.5
bs4: 4.5.0
html5lib: 0.999999999
sqlalchemy: None
pymysql: None
psycopg2: 2.7.5 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None
</details>

### PR DESCRIPTION
**Pull request by fujiaxiang**
- [ ] closes #26412
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by fujiaxiang** [comment]
There are two reasons for updating an old test case `test_loc_setitem_empty_append_raises` in `test_loc.py`:

1. To raise the same error as `.loc.__getitem__` when give slice does not exist

1. The current error says `cannot copy sequence with size 2 to array axis with dimension 0`. This is not in line with what the user is trying to do: to add 2 new rows or columns. The error should be a TypeError because we don't current support (not yet implemented) using integer slice to add new rows / columns.

Hope that sounds convincing.
If not, let me know.

**Comment by fujiaxiang** [comment]
This change has a huge impact on existing test cases. Let me investigate further.

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -1036,6 +1036,7 @@ Indexing
 - :meth:`MultiIndex.get_loc` can't find missing values when input includes missing values (:issue:`19132`)
 - Bug in :meth:`Series.__setitem__` incorrectly assigning values with boolean indexer when the length of new data matches the number of ``True`` values and new data is not a ``Series`` or an ``np.array`` (:issue:`30567`)
 - Bug in indexing with a :class:`PeriodIndex` incorrectly accepting integers representing years, use e.g. ``ser.loc["2007"]`` instead of ``ser.loc[2007]`` (:issue:`30763`)
+- Bug in :meth:`DataFrame.loc.__setitem__` and :meth:`Series.loc.__setitem__` not raising error when given integer slice does not exist in index (:issue:`26412`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v1.0.0.rst]:
```suggestion
- Bug in :meth:`DataFrame.loc.__setitem__` and :meth:`Series.loc.__setitem__` not raising error when given integer slice that do not exist in index (:issue:`26412`)
```

File: pandas/tests/indexing/test_loc.py [hunk 2]
```diff
@@ -985,6 +985,19 @@ def test_loc_setitem_float_intindex():
     tm.assert_frame_equal(result, expected)
 
 
+@pytest.mark.parametrize("test_series", [False, True])
+def test_loc_setitem_missing_integer_slice(test_series):
+    # GH 26412
+    df = DataFrame({"col1": [1, 2, 3]}, index=["a", "b", "c"])
+    msg = "cannot do slice indexing on .* with these indexers"
+    with pytest.raises(TypeError, match=msg):
```
Comment (gfyoung) [inline pandas/tests/indexing/test_loc.py]: Newline above this one

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -2922,6 +2922,9 @@ def is_int(v):
                     self._validate_indexer("slice", key.step, kind),
                 )
 
+        if kind == "loc" and not is_null_slicer:
+            return self.slice_indexer(key.start, key.stop, key.step, kind=kind)
+
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py]: is any of the validation on L2931-2948 relevant?  can some or all of it be removed?

This likely affects some tests, see how this is handled in #31840.

Comment (fujiaxiang) [reply pandas/core/indexes/base.py]: hey sorry I haven't had time to work on this.
will close this PR in favor of #31840.

**Comment by fujiaxiang** [comment]
closing this PR in favor of #31840.

**Pull request closed without merging**